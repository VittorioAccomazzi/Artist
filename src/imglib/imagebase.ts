import { textChangeRangeIsUnchanged } from "typescript";


interface TypedArray {
    readonly length : number;
    [n: number]: number;
    buffer : ArrayBuffer;
}

interface TypedArrayConstructor<T extends TypedArray> {
    new (buffer: ArrayBuffer, byteOffset?: number, length?: number): T;
}

export type ImageUint8  = ImageBase<Uint8Array,Uint8ArrayConstructor>
export type ImageUint16 = ImageBase<Uint16Array,Uint16ArrayConstructor>
export type ImageFloat32= ImageBase<Float32Array,Float32ArrayConstructor>

export class ImageFactory {

    static Uint8( width : number, height : number ) : ImageUint8 {
        return new ImageBase<Uint8Array,Uint8ArrayConstructor>(Uint8Array.BYTES_PER_ELEMENT, Uint8Array, width, height)
    }
    static Uint16( width : number, height : number ) : ImageUint16{
        return (new ImageBase<Uint16Array,Uint16ArrayConstructor>(Uint16Array.BYTES_PER_ELEMENT, Uint16Array, width, height)) 
    }
    static Float32( width : number, height : number ): ImageFloat32{
        return new ImageBase<Float32Array,Float32ArrayConstructor>(Float32Array.BYTES_PER_ELEMENT, Float32Array, width, height)
    }
}


class ImageBase<T extends TypedArray, C extends TypedArrayConstructor<T>> {

    private buffer : T
    private pixels : T [] 

    constructor( elementSize : number, constructor : C, width : number, height : number ) {
        let nPixels= width * height;
        let bytes  = nPixels * elementSize
        let buffer= new ArrayBuffer(bytes);
        this.pixels= Array.from(Array(height), (e,i)=> new constructor (buffer, width*elementSize*i, width) ) 
        this.buffer= new constructor(buffer, 0, nPixels)
    }


    /**
     * get type of pixel stored.
     */
    get imageType(): string {
        return this.pixels[0].constructor.name.replace('Array','')
    }

    /**
     * returns the current list of pixels.
     */
    get imagePixels(): T {
        return this.buffer 
    }

    /**
     *  get the image buffer
     */
    get imageBuffer() : ArrayBuffer {
        return this.buffer.buffer
    }

    /**
     * get image height
     */
    get height() : number {
        return this.pixels.length
    }

    /**
     * return image height
     */
    get width() : number {
        return this.pixels[0].length
    }

    /**
     * get a pixel value.
     * @param x row of the pixel queried
     * @param y column of the pixel queried
     */
    get( x: number, y: number ) : number {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates queried: image size ${this.width}x${this.height} pixel ${x},${y}`)
        return this.pixels[y][x]
    }

    /**
     * get a row of pixels
     * @param y row of pixels to select.
     */
    getRow(y:number) : T {
        if( y<0 || y>this.height )throw new Error (`Invalid y coordinate queried : image height ${this.height} y ${y}`)
        return this.pixels[y]
    }

    /**
     * 
     * @param x row pixel to set
     * @param y column pixel to se
     * @param value  value to set
     */
    set(x:number, y:number, value : number) : void {
        if( !this.checkBoundary(x,y)) throw new Error (`Invalid image coordinates set: image size ${this.width}x${this.height} pixel ${x},${y}`)
        this.pixels[y][x]= value
    }

    /**
     * check that the value passed are valid image coordinates.
     * @param x row
     * @param y column
     */
    private checkBoundary(x: number, y : number ) : boolean {
        return x>=0 && x < this.width && y>=0 && y<this.height
    }

}