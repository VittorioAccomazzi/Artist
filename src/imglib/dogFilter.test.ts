import {hash, getCanvases, toSeqCanvas, dumpImage, dumpCanvas, toNodeCanvas} from './testutils'
import {ImageFactory} from './imagebase'
import CanvasUtils from './canvasUtils'
import DifferenceOfGaussian from './dogFilter'

test('shall detect edges', async ()=>{

    let width = 320
    let height= 281
    let int1 = 100
    let int2 = 200
    let image = ImageFactory.Uint8(width,height)

    // first intensity
    image.imagePixels.forEach((v,i)=>image.imagePixels[i]=int1)

    // second intensity
    for( let y=Math.floor(height*0.3); y<Math.floor(height*0.8); y++){
        for( let x=Math.floor(width*0.4); x<Math.floor(width*0.8); x++){
            image.set(x,y,int2)
        }
    }

    let sSigma = 3
    let lSigma = Math.sqrt(1.6) * sSigma
    let sensitivity = 0.98
    let sharpness = 1
    let dogImage = DifferenceOfGaussian.Run( image, sSigma, lSigma, sensitivity, sharpness)
    let scaledDog= dogImage.convertTo('Uint8', 254,0)
    let hsh = await hash(scaledDog)
    expect(hsh).toMatchSnapshot()

    // await dumpImage(scaledDog,`dog - 3, 0.98, 1 `)
})

test('shall detect edges on Luminosity component ', async()=>{
    for await ( const [canvas,ctx,name] of getCanvases() ) {
        const [ inL ] = CanvasUtils.toLab(toSeqCanvas(canvas)) 

        let sSigma = 2
        let lSigma = 1.6 * sSigma
        let sensitivity = 0.98
        let sharpness = 1
        let dogImage = DifferenceOfGaussian.Run( inL, sSigma, lSigma, sensitivity, sharpness)
        let scaledDog= dogImage.convertTo('Uint8', 254,0)
        let hsh = await hash(scaledDog)
        expect(hsh).toMatchSnapshot()
    
         //await dumpImage(scaledDog, `dog edges on ${name}`)
    }
})