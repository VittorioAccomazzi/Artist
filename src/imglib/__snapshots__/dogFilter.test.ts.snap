// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shall detect edges 1`] = `"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffc0000000000001ffffffffffffffffffc0000000000000ffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff9ffffffffffffcffffffffffffffffff80000000000000ffffffffffffffffffc0000000000001ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"`;

exports[`shall detect edges on Luminosity component  1`] = `"fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffffffffffffffffff7fffd012000fffffffffffffffffffe620130000000fffffffffffffffffff0000000040000fffffffffffffff3ff80000000008000fffffffffff8f8000000000400000000fffffffffff8c0000000100000000000fffffffffff8003d8000004000000800fffffffffff0000000000000000004fffffffffffff0000000048ffe71fffff8ffffffff800000000797ffffffe80000ffc0000000000007fffffff9ffffffff0000000fffffffffffffffffffffdf9f3fffffff7ffffffffffffffffffff800fffffffffffffffffe7ffffffffffffffffffffffffffefffffffffffffffffffffffffeffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdffffffffffffffffffffffffffffffdfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffdffffffbfffffffff7fffffeffffdfffdffffdfffffffffffffffffbf3fffffffffffffbfffe3ffffffffffbff3fffffbfffffffffff1fffffffffdffffffffffffdfe3ffffffffe7ffc7ffffffffffdfffdfe3fff7fffc84ffffffffffe0fd8ffe1ffeffff3ff3c0f847fffffffffffffffffffceffe3ff09fffff99dfffffff37c7f03fcbfeffe01ff807fffffffffffffff17fffdf001fe4f3ff0404ffefff3fe27f7fffd0383033f0270e04fc9bfffeffff32ffe1c073ff906ef83ffc1ffff78e1ff7f1ffc07e03c071f83cfc7ffffe0fe3ffff87f8000ee0503f0007e180187c70e05fffc017fff879803ff385fffc7030ffebfe3f9ff004020e07801003f01ee3c789bfff9a880033f0f686ffe1fffffcdf2001fecff7dffece0f8ff83ffffe446a3f9f8380100002800fd3e7fefffe108bf900720fe207fefeacf5240c7f1c7a0bfc01773fe3fffbd883fffe0fff1e7099cc7ef402c000191f5f00cffffff9efff3ac8e703c400184084c0603ffef9e3f87e7eb3c3fd1fefcff8f87ff1dff0ff0061d8630f0000083067fc73f3e7fef4e7f1ff8f000ffffbfffb383801efdfcc2f0180007f9e991e0377f9fffe77bdfcc001c07c10702020e1c3cc1fffffffe34003804a15ff732b50e00191c3f9dff8f0a380be5f31000781e2e006d17ffbffe3c10072900024529fd20017fdfff181201680007f9b387c00126fe1fe0ffe78c1feb100459c039201f218fff951c7f8207900181800003c79ff5fd30fffffffc8ff60801f473840107fc96fe45e1c22ff30fcf999fe064043e43cffffff513d4f3f4000000600c1c83c183f8fff399f89724b183a0c332000fe3cbfe6fffff08ff8d002100117c001f0ec08a7ffc07eff481d80719e0031fc07fdffffee76c0a70183a49db38e0fa2f037dfc30237928ce01710a9020180069fc7c1e9fcefe7ffc1f002818002fe241000f6081c28fef000007c02216003df867e7ba1c61cc0361e11c3877e21f3f6c5f2cf780d9860fff038677c1c024e39c00fceec0c0c0701001bfc190300e98c280003f67fa4bfc420600160c83f3f81a03de9ff0e34abf40181010301f3fff1a4000cf0cf7083"`;

exports[`shall detect edges on Luminosity component  2`] = `"ffffffffffffffffffffffffff9fffffffffffffffffffffffffffffff9ffffcffffffffffffffffffffffffffcffff8ffffffffffffffffffffffffffe7fff1ffffffffffffffffffffffffffe7ffe3fffffffffffffffffffffffffff3ffc7fffffffffffffffffffffffffff3ff87fffffffffffffffffffffffffff9ff1ffffffffffffffffffffffffffffcfe3ffffffffffffffffffffffffffffcfc7fffffffffffffff8ffffffffffffe78ffffffffffffffffe1fffffffffffe71fffffffffffffffffc3fffffffffff23ffffffffffffffffff9fffffffffff87ffffffffffffffffffc7ffffffffff8ffffffffffffffffffff1ffffffffff9ffffbffffffff7ffffff8ffffffffff1f7ef3ffffffff7ffffffe7ffffffffe3ffcf3ffffffffffffffff1ffffffffc7ef8ffffffffffffffffffcffffffffcbff1ffffffffffffffffffc7fffffff9ffb3fffffffffffffffcffe3fff3fff1f3e7fffffffffffffff9fff3fff3fff3f7cfffffffffffffffe3fff8fff1ffe7cfdfffffffffffffffcffff87ff1ffc7cb3fffffffffffffff1ffffe3ff0ffcfd63ffffffffe1ffffc7fffff3ff0ff9fe47ffffffffe1ffff8ffffff1ff2091ff8fffffffffebffff3ffffff9ff21f3ff1ffffff7ffcbfffc7ffffff8ff03f3ff1fffffffffcfffb9fffffffcff09f3ff3fffffffffdbfff3fffffffc7f13f3f67fffffffffdb7fe7fffffffe3c07f3ec7fffffffff97ffcffffffdff38dff3fcffffffffff9fff9efffffbfff3fff3f8ffffffffff9ffffcffffffffcffff3f9ffffffffffbffe7bffffffffdffff3fbffffffffffbffef7ffffffffffffe7e7ffffffffffbffbcff7ffffffffffe7c7ffffffffffbff39fffffffffffff6747ffffffffff3fe7bffff7fffffffa4e9fffffffffff3fdf7ffff7ff7ffff74e9fffffffffff3f9fffe39ffffffff44d9fffffffffff3fb9ff80cfe1fffff08f3fffffffffff3ff3ff00f004ffffc11a3fffffffffff3d37ff0a4001fffe033e7fffffffffffbcfffe706903fff8043e7fffffffffff9ddffc888427ffe03cfe7fffffffffff9c9bfd04000fff87f9f0ffffffffffff9f3be640301fff17f1fcfffffffff7ffcd37f082407ffe0bc7f9ffffffffffffce7ffd0200ffff2b8ff9ffffffffffffe67de70181ffff871fb3fffffffff7ffe75f2b0303ffffc33ff3fffffffff7ffe677008107ffffc33fe7fffffffff7ffe6ff05020bffffc33fe7fffffffffffef0b804001bffffe17fe7fffffffff7fff0fc04203fffffe13fc7fffffffffffffbf100807bffffe13fcffffffffffffff8f00800fbfffff11e8ffffffffffffff9e13001fcfffff13f9fffffffffffffe0e00401f81fffc03f9ffffffffffffff0003003b307fe00bf9ffffffffffffff0000007a6a3fc789f3ffffffffffffcfa04000f801ff8189f3fffffffdff7f92804000f000ff0189f3fffffffdff7f92004001f00cfe0989e7fffffffdff7f04005040e48cfe098ae7fffffffdfffe18224043ce7ffe198a67fffffffdfffe700007c79fc3ff834a8ffffffffdff7fc0204f873ffffff7080ffffffff9fffe0001078e7fffffff484ffffffffdfffec000030cffffffff484ffffffffdffff00921308fffffffe489ffffffffdfffc41183219fffffffe0c9ffffffffdfffc011c2011fffffffe0c9ffffffff9ffac41202020fffffffc8c3ffffffff9ffa009032020fffffffcc43ffffffffdff0c00080040ffffff7c443ffffffff9ff3c00030240ffffff79667ffffffff9ffbcc0000380ffff8ff9e67ffffffffdff3910008312ffff81f1c67ffffffffffffc08041832fffffff3c67ffffffffbfffc09447272ffffffe7327ffffffffbffbc89686053ffffffe4b2ffffffffffff15342040d3ffffffcff2ffffffffffff0024114152fffffecff0ffffffffffff08260580727fc0009ed1ffffffffffff0805040033fff38f9ed3fffffffff7ff18040c04bb7fffff34d3fffffffff7ff0000244cbb7fffff3c43ffffffffe5ff2001209cfb7fffde6c43ffff7fffe7ff2a4d813cf93ffffc7c43ffff7fffe7ff36439835f90ffffc745bffff7fffe7ff2048d03d7f87fffd747bffff7ffff7ff0240d8657fa3fffdf47bfffe7ffce7ff10840e3d7ff01fe0fcf3fffe7ff0e5fe110400057f9ffff816f37ffeffe165fe020444897fdfffff0f7bfffeffc3e4ff000040097fdfffffc173fffcff8fecff092013015fcffffff073fffcff1ffcfe09000a017fcffffff833fffcfc3ffcfe180807017b4ffffffc33fffcfc7ffcfe1a000221dd4ffffffe33fffcfcff7cff11208189fc4fffffff33fffdf9ff3cfe11a00049c84fffffff97fffdf9ff3cfe0d01318838cfffffff87fffff3ff3cff8d0838003aefffffff87ffeff3ff3cff0e022800bcefffffffc7ffcfe7ff9cff2c040900bdcfffffffc7ffdfe7ff9cff9c24031073cfffffffe7ffffffff1cff1c2a218093cfffffffe7ffffffff9cff1100214027cfffffffe7ffffff7f9cff8048010027dfffffffe7ffffff7e9cff88400011ef9ffffffff1fffffffe9cff83d20005af9ffffffff0fffffcfe9cff8d5c1803bf3ffffffff2fffff9fe1efe411d8c8bbf3ffffffffbe03f9bff5efe011e4657be7ffffffff9f82f17f9defe1ded225e7c7ffffffff9fe6e27fffe7e9fc0009df8fffffffff9fe7c5ffd7e7e1f10401cc1fffffffffcff70dffa7e7e1812011e07fffffffffcff61fcf23e7f163208183ffffffffffcffe7fdfe8e7e36100017fffffffffffefffffdfece7e2f10240bfffffffffffe7fffdffccf7f1f000487fffffffffffe7fffffe54e7e0e00000ffffffffffffe7ffeff8f"`;
